// detritus-client-socket/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Package = Object.freeze({
    URL: 'https://github.com/detritusjs/client-socket',
    VERSION: '0.5.0',
});
function normalize(object) {
    for (let key in object) {
        object[key] = key;
    }
    return Object.freeze(object);
}
exports.ApiVersions = Object.freeze({
    GATEWAY: 6,
    MEDIA_GATEWAY: 5,
});
var CompressTypes;
(function (CompressTypes) {
    CompressTypes["NONE"] = "none";
    CompressTypes["PAYLOAD"] = "payload";
    CompressTypes["ZLIB"] = "zlib-stream";
})(CompressTypes = exports.CompressTypes || (exports.CompressTypes = {}));
exports.COMPRESS_TYPES = Object.freeze(Object.values(CompressTypes));
var CryptoModules;
(function (CryptoModules) {
    CryptoModules["LIBSODIUM_WRAPPERS"] = "libsodium-wrappers";
    CryptoModules["SODIUM"] = "sodium";
    CryptoModules["TWEETNACL"] = "tweetnacl";
})(CryptoModules = exports.CryptoModules || (exports.CryptoModules = {}));
exports.DEFAULT_SHARD_COUNT = 1;
exports.DEFAULT_SHARD_LAUNCH_DELAY = 5000;
exports.DEFAULT_VOICE_TIMEOUT = 30000;
var EncodingTypes;
(function (EncodingTypes) {
    EncodingTypes["ETF"] = "etf";
    EncodingTypes["JSON"] = "json";
})(EncodingTypes = exports.EncodingTypes || (exports.EncodingTypes = {}));
var GatewayActivityActionTypes;
(function (GatewayActivityActionTypes) {
    GatewayActivityActionTypes[GatewayActivityActionTypes["JOIN"] = 1] = "JOIN";
    GatewayActivityActionTypes[GatewayActivityActionTypes["SPECTATE"] = 2] = "SPECTATE";
    GatewayActivityActionTypes[GatewayActivityActionTypes["LISTEN"] = 3] = "LISTEN";
    GatewayActivityActionTypes[GatewayActivityActionTypes["WATCH"] = 4] = "WATCH";
    GatewayActivityActionTypes[GatewayActivityActionTypes["JOIN_REQUEST"] = 5] = "JOIN_REQUEST";
})(GatewayActivityActionTypes = exports.GatewayActivityActionTypes || (exports.GatewayActivityActionTypes = {}));
var GatewayActivityFlags;
(function (GatewayActivityFlags) {
    GatewayActivityFlags[GatewayActivityFlags["INSTANCE"] = 1] = "INSTANCE";
    GatewayActivityFlags[GatewayActivityFlags["JOIN"] = 2] = "JOIN";
    GatewayActivityFlags[GatewayActivityFlags["SPECTATE"] = 4] = "SPECTATE";
    GatewayActivityFlags[GatewayActivityFlags["JOIN_REQUEST"] = 8] = "JOIN_REQUEST";
    GatewayActivityFlags[GatewayActivityFlags["SYNC"] = 16] = "SYNC";
    GatewayActivityFlags[GatewayActivityFlags["PLAY"] = 32] = "PLAY";
})(GatewayActivityFlags = exports.GatewayActivityFlags || (exports.GatewayActivityFlags = {}));
var GatewayActivityTypes;
(function (GatewayActivityTypes) {
    GatewayActivityTypes[GatewayActivityTypes["PLAYING"] = 0] = "PLAYING";
    GatewayActivityTypes[GatewayActivityTypes["STREAMING"] = 1] = "STREAMING";
    GatewayActivityTypes[GatewayActivityTypes["LISTENING"] = 2] = "LISTENING";
    GatewayActivityTypes[GatewayActivityTypes["WATCHING"] = 3] = "WATCHING";
    GatewayActivityTypes[GatewayActivityTypes["CUSTOM_STATUS"] = 4] = "CUSTOM_STATUS";
})(GatewayActivityTypes = exports.GatewayActivityTypes || (exports.GatewayActivityTypes = {}));
exports.GatewayDispatchEvents = normalize({
    READY: null,
    RESUMED: null,
    ACTIVITY_JOIN_INVITE: null,
    ACTIVITY_JOIN_REQUEST: null,
    ACTIVITY_START: null,
    BRAINTREE_POPUP_BRIDGE_CALLBACK: null,
    CALL_CREATE: null,
    CALL_DELETE: null,
    CALL_UPDATE: null,
    CHANNEL_CREATE: null,
    CHANNEL_DELETE: null,
    CHANNEL_UPDATE: null,
    CHANNEL_PINS_ACK: null,
    CHANNEL_PINS_UPDATE: null,
    CHANNEL_RECIPIENT_ADD: null,
    CHANNEL_RECIPIENT_REMOVE: null,
    ENTITLEMENT_CREATE: null,
    ENTITLEMENT_DELETE: null,
    ENTITLEMENT_UPDATE: null,
    FRIEND_SUGGESTION_CREATE: null,
    FRIEND_SUGGESTION_DELETE: null,
    GIFT_CODE_UPDATE: null,
    GUILD_BAN_ADD: null,
    GUILD_BAN_REMOVE: null,
    GUILD_CREATE: null,
    GUILD_DELETE: null,
    GUILD_UPDATE: null,
    GUILD_EMOJIS_UPDATE: null,
    GUILD_INTEGRATIONS_UPDATE: null,
    GUILD_MEMBER_ADD: null,
    GUILD_MEMBER_LIST_UPDATE: null,
    GUILD_MEMBER_REMOVE: null,
    GUILD_MEMBER_UPDATE: null,
    GUILD_MEMBERS_CHUNK: null,
    GUILD_ROLE_CREATE: null,
    GUILD_ROLE_DELETE: null,
    GUILD_ROLE_UPDATE: null,
    INVITE_CREATE: null,
    INVITE_DELETE: null,
    LIBRARY_APPLICATION_UPDATE: null,
    LOBBY_CREATE: null,
    LOBBY_DELETE: null,
    LOBBY_UPDATE: null,
    LOBBY_MEMBER_CONNECT: null,
    LOBBY_MEMBER_DISCONNECT: null,
    LOBBY_MEMBER_UPDATE: null,
    LOBBY_MESSAGE: null,
    LOBBY_VOICE_SERVER_UPDATE: null,
    LOBBY_VOICE_STATE_UPDATE: null,
    MESSAGE_ACK: null,
    MESSAGE_CREATE: null,
    MESSAGE_DELETE: null,
    MESSAGE_DELETE_BULK: null,
    MESSAGE_REACTION_ADD: null,
    MESSAGE_REACTION_REMOVE: null,
    MESSAGE_REACTION_REMOVE_ALL: null,
    MESSAGE_REACTION_REMOVE_EMOJI: null,
    MESSAGE_UPDATE: null,
    OAUTH2_TOKEN_REMOVE: null,
    PRESENCES_REPLACE: null,
    PRESENCE_UPDATE: null,
    RECENT_MENTION_DELETE: null,
    RELATIONSHIP_ADD: null,
    RELATIONSHIP_REMOVE: null,
    SESSIONS_REPLACE: null,
    STREAM_CREATE: null,
    STREAM_DELETE: null,
    STREAM_SERVER_UPDATE: null,
    STREAM_UPDATE: null,
    TYPING_START: null,
    USER_ACHIEVEMENT_UPDATE: null,
    USER_CONNECTIONS_UPDATE: null,
    USER_FEED_SETTINGS_UPDATE: null,
    USER_GUILD_SETTINGS_UPDATE: null,
    USER_NOTE_UPDATE: null,
    USER_PAYMENT_SOURCES_UPDATE: null,
    USER_PAYMENTS_UPDATE: null,
    USER_SETTINGS_UPDATE: null,
    USER_SUBSCRIPTIONS_UPDATE: null,
    USER_REQUIRED_ACTION_UPDATE: null,
    USER_UPDATE: null,
    VOICE_SERVER_UPDATE: null,
    VOICE_STATE_UPDATE: null,
    WEBHOOKS_UPDATE: null,
});
var GatewayIntents;
(function (GatewayIntents) {
    GatewayIntents[GatewayIntents["GUILDS"] = 1] = "GUILDS";
    GatewayIntents[GatewayIntents["GUILD_MEMBERS"] = 2] = "GUILD_MEMBERS";
    GatewayIntents[GatewayIntents["GUILD_BANS"] = 4] = "GUILD_BANS";
    GatewayIntents[GatewayIntents["GUILD_EMOJIS"] = 8] = "GUILD_EMOJIS";
    GatewayIntents[GatewayIntents["GUILD_INTEGRATIONS"] = 16] = "GUILD_INTEGRATIONS";
    GatewayIntents[GatewayIntents["GUILD_WEBHOOKS"] = 32] = "GUILD_WEBHOOKS";
    GatewayIntents[GatewayIntents["GUILD_INVITES"] = 64] = "GUILD_INVITES";
    GatewayIntents[GatewayIntents["GUILD_VOICE_STATES"] = 128] = "GUILD_VOICE_STATES";
    GatewayIntents[GatewayIntents["GUILD_PRESENCES"] = 256] = "GUILD_PRESENCES";
    GatewayIntents[GatewayIntents["GUILD_MESSAGES"] = 512] = "GUILD_MESSAGES";
    GatewayIntents[GatewayIntents["GUILD_MESSAGE_REACTIONS"] = 1024] = "GUILD_MESSAGE_REACTIONS";
    GatewayIntents[GatewayIntents["GUILD_MESSAGE_TYPING"] = 2048] = "GUILD_MESSAGE_TYPING";
    GatewayIntents[GatewayIntents["DIRECT_MESSAGES"] = 4096] = "DIRECT_MESSAGES";
    GatewayIntents[GatewayIntents["DIRECT_MESSAGE_REACTIONS"] = 8192] = "DIRECT_MESSAGE_REACTIONS";
    GatewayIntents[GatewayIntents["DIRECT_MESSAGE_TYPING"] = 16384] = "DIRECT_MESSAGE_TYPING";
})(GatewayIntents = exports.GatewayIntents || (exports.GatewayIntents = {}));
var GatewayOpCodes;
(function (GatewayOpCodes) {
    GatewayOpCodes[GatewayOpCodes["DISPATCH"] = 0] = "DISPATCH";
    GatewayOpCodes[GatewayOpCodes["HEARTBEAT"] = 1] = "HEARTBEAT";
    GatewayOpCodes[GatewayOpCodes["IDENTIFY"] = 2] = "IDENTIFY";
    GatewayOpCodes[GatewayOpCodes["PRESENCE_UPDATE"] = 3] = "PRESENCE_UPDATE";
    GatewayOpCodes[GatewayOpCodes["VOICE_STATE_UPDATE"] = 4] = "VOICE_STATE_UPDATE";
    GatewayOpCodes[GatewayOpCodes["VOICE_SERVER_PING"] = 5] = "VOICE_SERVER_PING";
    GatewayOpCodes[GatewayOpCodes["RESUME"] = 6] = "RESUME";
    GatewayOpCodes[GatewayOpCodes["RECONNECT"] = 7] = "RECONNECT";
    GatewayOpCodes[GatewayOpCodes["REQUEST_GUILD_MEMBERS"] = 8] = "REQUEST_GUILD_MEMBERS";
    GatewayOpCodes[GatewayOpCodes["INVALID_SESSION"] = 9] = "INVALID_SESSION";
    GatewayOpCodes[GatewayOpCodes["HELLO"] = 10] = "HELLO";
    GatewayOpCodes[GatewayOpCodes["HEARTBEAT_ACK"] = 11] = "HEARTBEAT_ACK";
    GatewayOpCodes[GatewayOpCodes["SYNC_GUILD"] = 12] = "SYNC_GUILD";
    GatewayOpCodes[GatewayOpCodes["CALL_CONNECT"] = 13] = "CALL_CONNECT";
    GatewayOpCodes[GatewayOpCodes["GUILD_SUBSCRIPTIONS"] = 14] = "GUILD_SUBSCRIPTIONS";
    GatewayOpCodes[GatewayOpCodes["LOBBY_CONNECT"] = 15] = "LOBBY_CONNECT";
    GatewayOpCodes[GatewayOpCodes["LOBBY_DISCONNECT"] = 16] = "LOBBY_DISCONNECT";
    GatewayOpCodes[GatewayOpCodes["LOBBY_VOICE_STATES_UPDATE"] = 17] = "LOBBY_VOICE_STATES_UPDATE";
    GatewayOpCodes[GatewayOpCodes["STREAM_CREATE"] = 18] = "STREAM_CREATE";
    GatewayOpCodes[GatewayOpCodes["STREAM_DELETE"] = 19] = "STREAM_DELETE";
    GatewayOpCodes[GatewayOpCodes["STREAM_WATCH"] = 20] = "STREAM_WATCH";
    GatewayOpCodes[GatewayOpCodes["STREAM_PING"] = 21] = "STREAM_PING";
    GatewayOpCodes[GatewayOpCodes["STREAM_SET_PAUSED"] = 22] = "STREAM_SET_PAUSED";
})(GatewayOpCodes = exports.GatewayOpCodes || (exports.GatewayOpCodes = {}));
var GatewayPresenceStatuses;
(function (GatewayPresenceStatuses) {
    GatewayPresenceStatuses["ONLINE"] = "online";
    GatewayPresenceStatuses["DND"] = "dnd";
    GatewayPresenceStatuses["IDLE"] = "idle";
    GatewayPresenceStatuses["INVISIBLE"] = "invisible";
    GatewayPresenceStatuses["OFFLINE"] = "offline";
})(GatewayPresenceStatuses = exports.GatewayPresenceStatuses || (exports.GatewayPresenceStatuses = {}));
exports.MaxNumbers = Object.freeze({
    UINT8: 0xFF,
    UINT16: 0xFFFF,
    UINT32: 0xFFFFFFFF,
});
var MediaCodecTypes;
(function (MediaCodecTypes) {
    MediaCodecTypes["AUDIO"] = "audio";
    MediaCodecTypes["VIDEO"] = "video";
})(MediaCodecTypes = exports.MediaCodecTypes || (exports.MediaCodecTypes = {}));
var MediaCodecs;
(function (MediaCodecs) {
    MediaCodecs["OPUS"] = "opus";
    MediaCodecs["H264"] = "H264";
    MediaCodecs["VP8"] = "VP8";
    MediaCodecs["VP9"] = "VP9";
    MediaCodecs["RTX"] = "rtx";
})(MediaCodecs = exports.MediaCodecs || (exports.MediaCodecs = {}));
exports.MEDIA_CODECS_AUDIO = [
    MediaCodecs.OPUS,
];
exports.MEDIA_CODECS_VIDEO = [
    MediaCodecs.VP8,
    MediaCodecs.VP9,
    MediaCodecs.H264,
];
var MediaEncryptionModes;
(function (MediaEncryptionModes) {
    MediaEncryptionModes["XSALSA20_POLY1305_LITE"] = "xsalsa20_poly1305_lite";
    MediaEncryptionModes["XSALSA20_POLY1305_SUFFIX"] = "xsalsa20_poly1305_suffix";
    MediaEncryptionModes["XSALSA20_POLY1305"] = "xsalsa20_poly1305";
})(MediaEncryptionModes = exports.MediaEncryptionModes || (exports.MediaEncryptionModes = {}));
exports.MEDIA_ENCRYPTION_MODES = Object.freeze(Object.values(MediaEncryptionModes));
var MediaOpCodes;
(function (MediaOpCodes) {
    MediaOpCodes[MediaOpCodes["IDENTIFY"] = 0] = "IDENTIFY";
    MediaOpCodes[MediaOpCodes["SELECT_PROTOCOL"] = 1] = "SELECT_PROTOCOL";
    MediaOpCodes[MediaOpCodes["READY"] = 2] = "READY";
    MediaOpCodes[MediaOpCodes["HEARTBEAT"] = 3] = "HEARTBEAT";
    MediaOpCodes[MediaOpCodes["SELECT_PROTOCOL_ACK"] = 4] = "SELECT_PROTOCOL_ACK";
    MediaOpCodes[MediaOpCodes["SPEAKING"] = 5] = "SPEAKING";
    MediaOpCodes[MediaOpCodes["HEARTBEAT_ACK"] = 6] = "HEARTBEAT_ACK";
    MediaOpCodes[MediaOpCodes["RESUME"] = 7] = "RESUME";
    MediaOpCodes[MediaOpCodes["HELLO"] = 8] = "HELLO";
    MediaOpCodes[MediaOpCodes["RESUMED"] = 9] = "RESUMED";
    MediaOpCodes[MediaOpCodes["SIGNAL"] = 10] = "SIGNAL";
    MediaOpCodes[MediaOpCodes["CLIENT_CONNECT"] = 12] = "CLIENT_CONNECT";
    MediaOpCodes[MediaOpCodes["CLIENT_DISCONNECT"] = 13] = "CLIENT_DISCONNECT";
    MediaOpCodes[MediaOpCodes["SESSION_UPDATE"] = 14] = "SESSION_UPDATE";
    MediaOpCodes[MediaOpCodes["VIDEO_SINK_WANTS"] = 15] = "VIDEO_SINK_WANTS";
})(MediaOpCodes = exports.MediaOpCodes || (exports.MediaOpCodes = {}));
var MediaProtocols;
(function (MediaProtocols) {
    MediaProtocols["UDP"] = "udp";
    MediaProtocols["WEBRTC"] = "webrtc";
})(MediaProtocols = exports.MediaProtocols || (exports.MediaProtocols = {}));
exports.MEDIA_PROTOCOLS = Object.freeze(Object.values(MediaProtocols));
exports.MediaReceivedVideoQuality = Object.freeze({
    OFF: 'off',
    FULL: 'full',
});
exports.MediaSilencePacket = [0xF8, 0xFF, 0xFE];
var MediaSpeakingFlags;
(function (MediaSpeakingFlags) {
    MediaSpeakingFlags[MediaSpeakingFlags["NONE"] = 0] = "NONE";
    MediaSpeakingFlags[MediaSpeakingFlags["VOICE"] = 1] = "VOICE";
    MediaSpeakingFlags[MediaSpeakingFlags["SOUNDSHARE"] = 2] = "SOUNDSHARE";
    MediaSpeakingFlags[MediaSpeakingFlags["PRIORITY"] = 4] = "PRIORITY";
})(MediaSpeakingFlags = exports.MediaSpeakingFlags || (exports.MediaSpeakingFlags = {}));
var MediaSSRCTypes;
(function (MediaSSRCTypes) {
    MediaSSRCTypes["AUDIO"] = "audio";
    MediaSSRCTypes["VIDEO"] = "video";
})(MediaSSRCTypes = exports.MediaSSRCTypes || (exports.MediaSSRCTypes = {}));
var SocketEvents;
(function (SocketEvents) {
    SocketEvents["CLOSE"] = "close";
    SocketEvents["KILLED"] = "killed";
    SocketEvents["LOG"] = "log";
    SocketEvents["OPEN"] = "open";
    SocketEvents["PACKET"] = "packet";
    SocketEvents["READY"] = "ready";
    SocketEvents["RECONNECTING"] = "reconnecting";
    SocketEvents["SOCKET"] = "socket";
    SocketEvents["STATE"] = "state";
    SocketEvents["TRANSPORT"] = "transport";
    SocketEvents["TRANSPORT_READY"] = "transportReady";
    SocketEvents["WARN"] = "warn";
})(SocketEvents = exports.SocketEvents || (exports.SocketEvents = {}));
var SocketEventsBase;
(function (SocketEventsBase) {
    SocketEventsBase["CLOSE"] = "close";
    SocketEventsBase["ERROR"] = "error";
    SocketEventsBase["MESSAGE"] = "messsage";
    SocketEventsBase["OPEN"] = "open";
    SocketEventsBase["PING"] = "ping";
    SocketEventsBase["PONG"] = "pong";
})(SocketEventsBase = exports.SocketEventsBase || (exports.SocketEventsBase = {}));
var SocketCloseCodes;
(function (SocketCloseCodes) {
    SocketCloseCodes[SocketCloseCodes["NORMAL"] = 1000] = "NORMAL";
    SocketCloseCodes[SocketCloseCodes["GOING_AWAY"] = 1001] = "GOING_AWAY";
    SocketCloseCodes[SocketCloseCodes["PROTOCOL_ERROR"] = 1002] = "PROTOCOL_ERROR";
    SocketCloseCodes[SocketCloseCodes["UNSUPPORTED_DATA"] = 1003] = "UNSUPPORTED_DATA";
    SocketCloseCodes[SocketCloseCodes["ABNORMAL_CLOSURE"] = 1006] = "ABNORMAL_CLOSURE";
    SocketCloseCodes[SocketCloseCodes["INVALID_FRAME"] = 1007] = "INVALID_FRAME";
    SocketCloseCodes[SocketCloseCodes["POLICY_VIOLATION"] = 1008] = "POLICY_VIOLATION";
    SocketCloseCodes[SocketCloseCodes["MESSAGE_TOO_BIG"] = 1009] = "MESSAGE_TOO_BIG";
    SocketCloseCodes[SocketCloseCodes["MISSING_EXTENSION"] = 1010] = "MISSING_EXTENSION";
    SocketCloseCodes[SocketCloseCodes["INTERNAL_ERROR"] = 1011] = "INTERNAL_ERROR";
    SocketCloseCodes[SocketCloseCodes["SERVICE_RESTART"] = 1012] = "SERVICE_RESTART";
    SocketCloseCodes[SocketCloseCodes["TRY_AGAIN_LATER"] = 1013] = "TRY_AGAIN_LATER";
    SocketCloseCodes[SocketCloseCodes["BAD_GATEWAY"] = 1014] = "BAD_GATEWAY";
})(SocketCloseCodes = exports.SocketCloseCodes || (exports.SocketCloseCodes = {}));
var SocketInternalCloseCodes;
(function (SocketInternalCloseCodes) {
    SocketInternalCloseCodes[SocketInternalCloseCodes["INVALID_DATA"] = 4800] = "INVALID_DATA";
    SocketInternalCloseCodes[SocketInternalCloseCodes["RECONNECTING"] = 4801] = "RECONNECTING";
    SocketInternalCloseCodes[SocketInternalCloseCodes["HEARTBEAT_ACK"] = 4802] = "HEARTBEAT_ACK";
    SocketInternalCloseCodes[SocketInternalCloseCodes["HEARTBEAT_ACK_NONCE"] = 4803] = "HEARTBEAT_ACK_NONCE";
    SocketInternalCloseCodes[SocketInternalCloseCodes["OTHER_SOCKET_MESSAGE"] = 4804] = "OTHER_SOCKET_MESSAGE";
    SocketInternalCloseCodes[SocketInternalCloseCodes["OTHER_SOCKET_OPEN"] = 4805] = "OTHER_SOCKET_OPEN";
})(SocketInternalCloseCodes = exports.SocketInternalCloseCodes || (exports.SocketInternalCloseCodes = {}));
exports.SocketInternalCloseReasons = Object.freeze({
    [SocketInternalCloseCodes.INVALID_DATA]: 'Invalid data received, reconnecting',
    [SocketInternalCloseCodes.RECONNECTING]: 'Reconnecting',
    [SocketInternalCloseCodes.HEARTBEAT_ACK]: 'Heartbeat ACK never arrived',
    [SocketInternalCloseCodes.HEARTBEAT_ACK_NONCE]: 'Invalid nonce received by Heartbeat ACK',
    [SocketInternalCloseCodes.OTHER_SOCKET_MESSAGE]: 'Received message from not our current socket',
    [SocketInternalCloseCodes.OTHER_SOCKET_OPEN]: 'Received open from not our current socket',
});
var SocketGatewayCloseCodes;
(function (SocketGatewayCloseCodes) {
    SocketGatewayCloseCodes[SocketGatewayCloseCodes["UNKNOWN_ERROR"] = 4000] = "UNKNOWN_ERROR";
    SocketGatewayCloseCodes[SocketGatewayCloseCodes["UNKNOWN_OPCODE"] = 4001] = "UNKNOWN_OPCODE";
    SocketGatewayCloseCodes[SocketGatewayCloseCodes["DECODE_ERROR"] = 4002] = "DECODE_ERROR";
    SocketGatewayCloseCodes[SocketGatewayCloseCodes["NOT_AUTHENTICATED"] = 4003] = "NOT_AUTHENTICATED";
    SocketGatewayCloseCodes[SocketGatewayCloseCodes["AUTHENTICATION_FAILED"] = 4004] = "AUTHENTICATION_FAILED";
    SocketGatewayCloseCodes[SocketGatewayCloseCodes["ALREADY_AUTHENTICATED"] = 4005] = "ALREADY_AUTHENTICATED";
    SocketGatewayCloseCodes[SocketGatewayCloseCodes["INVALID_SEQUENCE"] = 4007] = "INVALID_SEQUENCE";
    SocketGatewayCloseCodes[SocketGatewayCloseCodes["RATE_LIMITED"] = 4008] = "RATE_LIMITED";
    SocketGatewayCloseCodes[SocketGatewayCloseCodes["SESSION_TIMEOUT"] = 4009] = "SESSION_TIMEOUT";
    SocketGatewayCloseCodes[SocketGatewayCloseCodes["INVALID_SHARD"] = 4010] = "INVALID_SHARD";
    SocketGatewayCloseCodes[SocketGatewayCloseCodes["SHARDING_REQUIRED"] = 4011] = "SHARDING_REQUIRED";
    SocketGatewayCloseCodes[SocketGatewayCloseCodes["INVALID_VERSION"] = 4012] = "INVALID_VERSION";
    SocketGatewayCloseCodes[SocketGatewayCloseCodes["INVALID_INTENTS"] = 4013] = "INVALID_INTENTS";
    SocketGatewayCloseCodes[SocketGatewayCloseCodes["DISALLOWED_INTENTS"] = 4014] = "DISALLOWED_INTENTS";
})(SocketGatewayCloseCodes = exports.SocketGatewayCloseCodes || (exports.SocketGatewayCloseCodes = {}));
;
var SocketMediaCloseCodes;
(function (SocketMediaCloseCodes) {
    SocketMediaCloseCodes[SocketMediaCloseCodes["UNKNOWN_ERROR"] = 4000] = "UNKNOWN_ERROR";
    SocketMediaCloseCodes[SocketMediaCloseCodes["UNKNOWN_OPCODE"] = 4001] = "UNKNOWN_OPCODE";
    SocketMediaCloseCodes[SocketMediaCloseCodes["DECODE_ERROR"] = 4002] = "DECODE_ERROR";
    SocketMediaCloseCodes[SocketMediaCloseCodes["NOT_AUTHENTICATED"] = 4003] = "NOT_AUTHENTICATED";
    SocketMediaCloseCodes[SocketMediaCloseCodes["AUTHENTICATION_FAILED"] = 4004] = "AUTHENTICATION_FAILED";
    SocketMediaCloseCodes[SocketMediaCloseCodes["ALREADY_AUTHENTICATED"] = 4005] = "ALREADY_AUTHENTICATED";
    SocketMediaCloseCodes[SocketMediaCloseCodes["SESSION_NO_LONGER_VALID"] = 4006] = "SESSION_NO_LONGER_VALID";
    SocketMediaCloseCodes[SocketMediaCloseCodes["SESSION_TIMEOUT"] = 4009] = "SESSION_TIMEOUT";
    SocketMediaCloseCodes[SocketMediaCloseCodes["SERVER_NOT_FOUND"] = 4011] = "SERVER_NOT_FOUND";
    SocketMediaCloseCodes[SocketMediaCloseCodes["UNKNOWN_PROTOCOL"] = 4012] = "UNKNOWN_PROTOCOL";
    SocketMediaCloseCodes[SocketMediaCloseCodes["DISCONNECTED"] = 4014] = "DISCONNECTED";
    SocketMediaCloseCodes[SocketMediaCloseCodes["VOICE_SERVER_CRASHED"] = 4015] = "VOICE_SERVER_CRASHED";
    SocketMediaCloseCodes[SocketMediaCloseCodes["UNKNOWN_ENCRYPTION_MODE"] = 4016] = "UNKNOWN_ENCRYPTION_MODE";
})(SocketMediaCloseCodes = exports.SocketMediaCloseCodes || (exports.SocketMediaCloseCodes = {}));
;
var SocketStates;
(function (SocketStates) {
    SocketStates["CLOSED"] = "CLOSED";
    SocketStates["CONNECTING"] = "CONNECTING";
    SocketStates["IDENTIFYING"] = "IDENTIFYING";
    SocketStates["OPEN"] = "OPEN";
    SocketStates["READY"] = "READY";
    SocketStates["RESUMING"] = "RESUMING";
})(SocketStates = exports.SocketStates || (exports.SocketStates = {}));
exports.RTP_HEADER_VERSION = 0x80;
var RTPPayloadTypes;
(function (RTPPayloadTypes) {
    RTPPayloadTypes[RTPPayloadTypes["OPUS"] = 120] = "OPUS";
    RTPPayloadTypes[RTPPayloadTypes["VP8"] = 101] = "VP8";
    RTPPayloadTypes[RTPPayloadTypes["VP9"] = 103] = "VP9";
    RTPPayloadTypes[RTPPayloadTypes["H264"] = 105] = "H264";
})(RTPPayloadTypes = exports.RTPPayloadTypes || (exports.RTPPayloadTypes = {}));
exports.RTP_PAYLOAD_TYPES = Object.freeze(Object.values(RTPPayloadTypes));
exports.RTCP_HEADER_VERSION = 0x80;
var RTCPPacketTypes;
(function (RTCPPacketTypes) {
    RTCPPacketTypes[RTCPPacketTypes["SENDER_REPORT"] = 200] = "SENDER_REPORT";
    RTCPPacketTypes[RTCPPacketTypes["RECEIVER_REPORT"] = 201] = "RECEIVER_REPORT";
    RTCPPacketTypes[RTCPPacketTypes["SOURCE_DESCRIPTION"] = 202] = "SOURCE_DESCRIPTION";
    RTCPPacketTypes[RTCPPacketTypes["BYE"] = 203] = "BYE";
    RTCPPacketTypes[RTCPPacketTypes["APP"] = 204] = "APP";
    RTCPPacketTypes[RTCPPacketTypes["RTPFB"] = 205] = "RTPFB";
    RTCPPacketTypes[RTCPPacketTypes["PSFB"] = 206] = "PSFB";
})(RTCPPacketTypes = exports.RTCPPacketTypes || (exports.RTCPPacketTypes = {}));
exports.RTCP_PACKET_TYPES = Object.freeze(Object.values(RTCPPacketTypes));
exports.RTPHeaderExtensionOneByte = Object.freeze({
    HEADER: [0xBE, 0xDE],
    LOCAL_IDENTIFER: 0xF,
});
exports.RTPHeaderExtensionTwoByte = Object.freeze({
    HEADER: [0x10, 0x00],
});
exports.ZLIB_SUFFIX = [0x0, 0x0, 0xff, 0xff];
